File: external/m2html/templates/frame/doxysearch.php
<?php
/******************************************************************************
 *
 * $Id:$
 *
 * Copyright (C) 1997-2003 by Dimitri van Heesch.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation under the terms of the GNU General Public License is hereby 
 * granted. No representations are made about the suitability of this software 
 * for any purpose. It is provided "as is" without express or implied warranty.
 * See the GNU General Public License for more details.
 *
 */

function readInt($file)
{
  $b1 = ord(fgetc($file)); $b2 = ord(fgetc($file));
  $b3 = ord(fgetc($file)); $b4 = ord(fgetc($file));
  return ($b1<<24)|($b2<<16)|($b3<<8)|$b4;
}

function readString($file)
{
  $result="";
  while (ord($c=fgetc($file))) $result.=$c;
  return $result;
}

function readHeader($file)
{
    $header =fgetc($file); $header.=fgetc($file);
    $header.=fgetc($file); $header.=fgetc($file);
    return $header;
}

function computeIndex($word)
{
  if (strlen($word)<2) return -1;
  // high char of the index
  $hi = ord($word{0});
  if ($hi==0) return -1;
  // low char of the index
  $lo = ord($word{1});
  if ($lo==0) return -1;
  // return index
  return $hi*256+$lo;
}

function search($file,$word,&$statsList)
{
  $index = computeIndex($word);
  if ($index!=-1) // found a valid index
  {
    fseek($file,$index*4+4); // 4 bytes per entry, skip header
    $index = readInt($file);
    if ($index) // found words matching first two characters
    {
      $start=sizeof($statsList);
      $count=$start;
      fseek($file,$index);
      $w = readString($file);
      while ($w)
      {
        $statIdx = readInt($file);
        if ($word==substr($w,0,strlen($word)))
        { // found word that matches (as substring)
          $statsList[$count++]=array(
              "word"=>$word,
              "match"=>$w,
              "index"=>$statIdx,
              "full"=>strlen($w)==strlen($word),
              "docs"=>array()
              );
        }
        $w = readString($file);
      }
      $totalFreq=0;
      for ($count=$start;$count<sizeof($statsList);$count++)
      {
        $statInfo = &$statsList[$count];
        fseek($file,$statInfo["index"]); 
        $numDocs = readInt($file);
        $docInfo = array();
        // read docs info + occurrence frequency of the word
        for ($i=0;$i<$numDocs;$i++)
        {
          $idx=readInt($file); 
          $freq=readInt($file); 
          $docInfo[$i]=array("idx"=>$idx,"freq"=>$freq,"rank"=>0.0);
          $totalFreq+=$freq;
          if ($statInfo["full"]) $totalfreq+=$freq;
        }
        // read name an url info for the doc
        for ($i=0;$i<$numDocs;$i++)
        {
          fseek($file,$docInfo[$i]["idx"]);
          $docInfo[$i]["name"]=readString($file);
          $docInfo[$i]["url"]=readString($file);
        }
        $statInfo["docs"]=$docInfo;
      }
      for ($count=$start;$count<sizeof($statsList);$count++)
      {
        $statInfo = &$statsList[$count];
        for ($i=0;$i<sizeof($statInfo["docs"]);$i++)
        {
          $docInfo = &$statInfo["docs"];
          // compute frequency rank of the word in each doc
          $statInfo["docs"][$i]["rank"]=
            (float)$docInfo[$i]["freq"]/$totalFreq;
        }
      }
    }
  }
  return $statsList;
}

function combine_results($results,&$docs)
{
  foreach ($results as $wordInfo)
  {
    $docsList = &$wordInfo["docs"];
    foreach ($docsList as $di)
    {
      $key=$di["url"];
      $rank=$di["rank"];
      if (in_array($key, array_keys($docs)))
      {
        $docs[$key]["rank"]+=$rank;
        $docs[$key]["rank"]*=2; // multiple matches increases rank 
      }
      else
      {
        $docs[$key] = array("url"=>$key,
            "name"=>$di["name"],
            "rank"=>$rank
            );
      }
      $docs[$key]["words"][] = array(
               "word"=>$wordInfo["word"],
               "match"=>$wordInfo["match"],
               "freq"=>$di["freq"]
               );
    }
  }
  return $docs;
}

function normalize_ranking(&$docs)
{
  $maxRank = 0.0000001;
  // compute maximal rank
  foreach ($docs as $doc) 
  {
    if ($doc["rank"]>$maxRank)
    {
      $maxRank=$doc["rank"];
    }
  }
  reset($docs);
  // normalize rankings
  while (list ($key, $val) = each ($docs)) 
  {
    $docs[$key]["rank"]*=100/$maxRank;
  }
}

function filter_results($docs,&$requiredWords,&$forbiddenWords)
{
  $filteredDocs=array();
  while (list ($key, $val) = each ($docs)) 
  {
    $words = &$docs[$key]["words"];
    $copy=1; // copy entry by default
    if (sizeof($requiredWords)>0)
    {
      foreach ($requiredWords as $reqWord)
      {
        $found=0;
        foreach ($words as $wordInfo)
        { 
          $found = $wordInfo["word"]==$reqWord;
          if ($found) break;
        }
        if (!$found) 
        {
          $copy=0; // document contains none of the required words
          break;
        }
      }
    }
    if (sizeof($forbiddenWords)>0)
    {
      foreach ($words as $wordInfo)
      {
        if (in_array($wordInfo["word"],$forbiddenWords))
        {
          $copy=0; // document contains a forbidden word
          break;
        }
      }
    }
    if ($copy) $filteredDocs[$key]=$docs[$key];
  }
  return $filteredDocs;
}

function compare_rank($a,$b)
{
  return ($a["rank"]>$b["rank"]) ? -1 : 1; 
}

function sort_results($docs,&$sorted)
{
  $sorted = $docs;
  usort($sorted,"compare_rank");
  return $sorted;
}

function report_results(&$docs)
{
  echo "<table cellspacing=\"2\">\n";
  echo "  <tr>\n";
  echo "    <td colspan=\"2\"><h2>Search Results</h2></td>\n";
  echo "  </tr>\n";
  $numDocs = sizeof($docs);
  if ($numDocs==0)
  {
    echo "  <tr>\n";
    echo "    <td colspan=\"2\">".matches_text(0)."</td>\n";
    echo "  </tr>\n";
  }
  else
  {
    echo "  <tr>\n";
    echo "    <td colspan=\"2\">".matches_text($numDocs);
    echo "\n";
    echo "    </td>\n";
    echo "  </tr>\n";
    $num=1;
    foreach ($docs as $doc)
    {
      echo "  <tr>\n";
      echo "    <td align=\"right\">$num.</td>";
      echo     "<td><a class=\"el\" href=\"".$doc["url"]."\">".$doc["name"]."</a></td>\n";
      echo "  <tr>\n";
      echo "    <td></td><td class=\"tiny\">Matches: ";
      foreach ($doc["words"] as $wordInfo)
      {
        $word = $wordInfo["word"];
        $matchRight = substr($wordInfo["match"],strlen($word));
        echo "<b>$word</b>$matchRight(".$wordInfo["freq"].") ";
      }
      echo "    </td>\n";
      echo "  </tr>\n";
      $num++;
    }
  }
  echo "</table>\n";
}

function matches_text($num)
{
  if ($num==0)
  {
    return 'Sorry, no documents matching your query.';
  }
  else if ($num==1)
  {
    return 'Found 1 document matching your query.';
  }
  else // $num>1
  {
    return 'Found '.$num.' documents matching your query. Showing best matches first.';
  }
}

function main($idxfile)
{
  if(strcmp('4.1.0', phpversion()) > 0) 
  {
    die("Error: PHP version 4.1.0 or above required!");
  }
  if (!($file=fopen($idxfile,"rb"))) 
  {
    die("Error: Search index file could NOT be opened!");
  }
  if (readHeader($file)!="DOXS")
  {
    die("Error: Header of index file is invalid!");
  }
  $query="";
  if (array_key_exists("query", $_GET))
  {
    $query=$_GET["query"];
  }
  $results = array();
  $requiredWords = array();
  $forbiddenWords = array();
  $foundWords = array();
  $word=strtolower(strtok($query," "));
  while ($word) // for each word in the search query
  {
    if (($word{0}=='+')) { $word=substr($word,1); $requiredWords[]=$word; }
    if (($word{0}=='-')) { $word=substr($word,1); $forbiddenWords[]=$word; }
    if (!in_array($word,$foundWords))
    {
      $foundWords[]=$word;
      search($file,$word,$results);
    }
    $word=strtolower(strtok(" "));
  }
  $docs = array();
  combine_results($results,$docs);
  // filter out documents with forbidden word or that do not contain
  // required words
  $filteredDocs = filter_results($docs,$requiredWords,$forbiddenWords);
  // normalize rankings so they are in the range [0-100]
  normalize_ranking($filteredDocs);
  // sort the results based on rank
  $sorted = array();
  sort_results($filteredDocs,$sorted);
  // report results to the user
  report_results($sorted);
  fclose($file);
}

?>

File: external/m2html/private/doxysearch.m
function result = doxysearch(query,filename)
%DOXYSEARCH Search a query in a 'search.idx' file
%  RESULT = DOXYSEARCH(QUERY,FILENAME) looks for request QUERY
%  in FILENAME (Doxygen search.idx format) and returns a list of
%  files responding to the request in RESULT.
%
%  See also DOXYREAD, DOXYWRITE

%  Copyright (C) 2004 Guillaume Flandin <Guillaume@artefact.tk>
%  $Revision: 1.1 $Date: 2004/05/05 14:33:55 $

%  This program is free software; you can redistribute it and/or
%  modify it under the terms of the GNU General Public License
%  as published by the Free Software Foundation; either version 2
%  of the License, or any later version.
% 
%  This program is distributed in the hope that it will be useful,
%  but WITHOUT ANY WARRANTY; without even the implied warranty of
%  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%  GNU General Public License for more details.
% 
%  You should have received a copy of the GNU General Public License
%  along with this program; if not, write to the Free Software
%  Foundation Inc, 59 Temple Pl. - Suite 330, Boston, MA 02111-1307, USA.

%  Suggestions for improvement and fixes are always welcome, although no
%  guarantee is made whether and when they will be implemented.
%  Send requests to <Guillaume@artefact.tk>

%  See <http://www.doxygen.org/> for more details.

error(nargchk(1,2,nargin));
if nargin == 1,
	filename = 'search.idx';
end

%- Open the search index file
[fid, errmsg] = fopen(filename,'r','ieee-be');
if fid == -1, error(errmsg); end

%- 4 byte header (DOXS)
header = char(fread(fid,4,'uchar'))';
if ~all(header == 'DOXS')
    error('[doxysearch] Header of index file is invalid!');
end

%- many thanks to <doxyread.m> and <doxysearch.php>
r = query;
requiredWords  = {};
forbiddenWords = {};
foundWords     = {};
res            = {};
while 1
    % extract each word of the query
    [t,r] = strtok(r);
    if isempty(t), break, end;
    if t(1) == '+'
        t = t(2:end); requiredWords{end+1} = t;
    elseif t(1) == '-'
        t = t(2:end); forbiddenWords{end+1} = t;
    end
    if ~ismember(t,foundWords)
        foundWords{end+1} = t;
        res = searchAgain(fid,t,res);
    end
end

%- Filter and sort results
docs = combineResults(res);
filtdocs = filterResults(docs,requiredWords,forbiddenWords);
filtdocs = normalizeResults(filtdocs);
res = sortResults(filtdocs);

%- 
if nargout
    result = res;
else
    for i=1:size(res,1)
        fprintf('   %d. %s - %s\n      ',i,res{i,1},res{i,2});
        for j=1:size(res{i,4},1)
            fprintf('%s ',res{i,4}{j,1});
        end
        fprintf('\n');
    end
end

%- Close the search index file
fclose(fid);

%===========================================================================
function res = searchAgain(fid, word,res)

	i = computeIndex(word);
	if i > 0
        
        fseek(fid,i*4+4,'bof'); % 4 bytes per entry, skip header
        start = size(res,1);
        idx = readInt(fid);
        
        if idx > 0
            
            fseek(fid,idx,'bof');
            statw = readString(fid);
		    while ~isempty(statw)
			    statidx  = readInt(fid);
                if length(statw) >= length(word) & ...
                    strcmp(statw(1:length(word)),word)
			        res{end+1,1} = statw;   % word
                    res{end,2}   = word;    % match
			        res{end,3}   = statidx; % index
                    res{end,4}   = (length(statw) == length(word)); % full
                    res{end,5}   = {};      % doc
                end
			    statw = readString(fid);
        	end
        
            totalfreq = 0;
            for j=start+1:size(res,1)
                fseek(fid,res{j,3},'bof');
                numdoc = readInt(fid);
                docinfo = {};
                for m=1:numdoc
			        docinfo{m,1} = readInt(fid); % idx
			        docinfo{m,2} = readInt(fid); % freq
                    docinfo{m,3} = 0;            % rank
                    totalfreq = totalfreq + docinfo{m,2};
                    if res{j,2}, 
                        totalfreq = totalfreq + docinfo{m,2};
                    end;
		        end
		        for m=1:numdoc
			        fseek(fid, docinfo{m,1}, 'bof');
			        docinfo{m,4} = readString(fid); % name
			        docinfo{m,5} = readString(fid); % url
		        end
                res{j,5} = docinfo;
            end
        
            for j=start+1:size(res,1)
                for m=1:size(res{j,5},1)
                    res{j,5}{m,3} = res{j,5}{m,2} / totalfreq;
                end
            end
            
        end % if idx > 0
        
	end % if i > 0

%===========================================================================
function docs = combineResults(result)

	docs = {};
	for i=1:size(result,1)
        for j=1:size(result{i,5},1)
            key = result{i,5}{j,5};
            rank = result{i,5}{j,3};
            if ~isempty(docs) & ismember(key,{docs{:,1}})
                l = find(ismember({docs{:,1}},key));
                docs{l,3} = docs{l,3} + rank;
                docs{l,3} = 2 * docs{l,3};
            else
                l = size(docs,1)+1;
                docs{l,1} = key; % key
                docs{l,2} = result{i,5}{j,4}; % name
                docs{l,3} = rank; % rank
                docs{l,4} = {}; %words
            end
            n = size(docs{l,4},1);
            docs{l,4}{n+1,1} = result{i,1}; % word
            docs{l,4}{n+1,2} = result{i,2}; % match
            docs{l,4}{n+1,3} = result{i,5}{j,2}; % freq
        end
	end

%===========================================================================
function filtdocs = filterResults(docs,requiredWords,forbiddenWords)

	filtdocs = {};
	for i=1:size(docs,1)
        words = docs{i,4};
        c = 1;
        j = size(words,1);
        % check required
        if ~isempty(requiredWords)
            found = 0;
            for k=1:j
                if ismember(words{k,1},requiredWords)
                    found = 1; 
                    break;  
                end
            end
            if ~found, c = 0; end
        end
        % check forbidden
        if ~isempty(forbiddenWords)
            for k=1:j
                if ismember(words{k,1},forbiddenWords)
                    c = 0;
                    break;
                end
            end
        end
        % keep it or not
        if c, 
            l = size(filtdocs,1)+1;
            filtdocs{l,1} = docs{i,1};
            filtdocs{l,2} = docs{i,2};
            filtdocs{l,3} = docs{i,3};
            filtdocs{l,4} = docs{i,4};
        end;
	end

%===========================================================================
function docs = normalizeResults(docs);

    m = max([docs{:,3}]);
    for i=1:size(docs,1)
        docs{i,3} = 100 * docs{i,3} / m;
    end

%===========================================================================
function result = sortResults(docs);

    [y, ind] = sort([docs{:,3}]);
    result = {};
    ind = fliplr(ind);
    for i=1:size(docs,1)
        result{i,1} = docs{ind(i),1};
        result{i,2} = docs{ind(i),2};
        result{i,3} = docs{ind(i),3};
        result{i,4} = docs{ind(i),4};
    end

%===========================================================================
function i = computeIndex(word)

    if length(word) < 2,
       i = -1;
    else
        i = double(word(1)) * 256 + double(word(2));
    end
    
%===========================================================================
function s = readString(fid)

	s = '';
	while 1
		w = fread(fid,1,'uchar');
		if w == 0, break; end
		s(end+1) = char(w);
	end

%===========================================================================
function i = readInt(fid)

	i = fread(fid,1,'uint32');